static/js/api.js

```diff
@@ -89,50 +89,115 @@ function saveUserLocation(loc) {
   return loc;
 }
 
-async function searchLocation(q, loc, kind = '') {
-  if (q.length <= 2) {
-    return { q, results: [] };
-  }
-  let url = `https://api-adresse.data.gouv.fr/search/?q=${encodeURIComponent(q)}&limit=5&type=${kind}`;
+function getMunicipalityApi(q) {
+  return `https://geo.api.gouv.fr/communes?nom=${encodeURIComponent(
+    q
+  )}&boost=population&fields=centre,codesPostaux&limit=5`;
+}
+
+function getAddressApi(q, loc) {
+  let url = `https://api-adresse.data.gouv.fr/search/?q=${encodeURIComponent(q)}&limit=5`;
   const { lat, lon } = loc || {};
   if (lat && lon) url += `&lat=${lat}&lon=${lon}`;
-  try {
-    const res = await fetch(url);
-    const { features } = await res.json();
-    const results = features.filter((
-      { properties: { score } }) => {
-      return (score > 0.4);
-    }).map(
-      ({ properties: { type, label, context, citycode, postcode, search_type, city, street,municipality }, geometry: { coordinates } }) => {
-        const text = type === "municipality" ? `${label} (${postcode})` : label;
-        context = ""  // Empty the context, which initially contains department, region
-
-        return {
-          id: "loc",
-          text,
-          context: context,
-          code: citycode,
-          lat: coordinates[1],
-          lon: coordinates[0],
-          search_type: type,
-          postcode: postcode,
-          street_name: street,
-          municipality: municipality,
-        };
-      }
-    );
+  return url;
+}
+
+// Global TODO
+// Move this to another FILE ?
+// Check in response: "id", context
+
+function buildResultFromAddressApi({
+  properties: { type, label, context, citycode, postcode, search_type, city, street, municipality },
+  geometry: { coordinates },
+}) {
+  context = ""; // Empty the context, which initially contains department, region
+
+  return {
+    id: "loc",
+    label,
+    context: context,
+    code: citycode,
+    lat: coordinates[1],
+    lon: coordinates[0],
+    search_type: type,
+    postcode: postcode,
+    street_name: street,
+    municipality: municipality,
+  };
+}
+
+
+function buildResultFromMunicipalityApi({ code, nom, centre, codesPostaux }) {
+  const text = `${nom} (${codesPostaux})`;
+  return {
+    id: "loc",
+    text,
+    context: "",
+    code: code,
+    lat: centre["coordinates"][1],
+    lon: centre["coordinates"][0],
+    search_type: "municipality",
+    postcode: codesPostaux,
+    street_name: "",
+    municipality: nom,
+  };
+}
+
+function flattenPostCodeFromResult(result, postcode) {
+  let clone = JSON.parse(JSON.stringify(result));
+  clone["postcode"] = postcode;
+  clone["text"] = `${result["municipality"]} (${postcode})`;
+  return clone;
+}
 
-    return { q, results };
-  } catch (err) {
-    // most likely a temporary network issue
-    console.warn(`error while searching location: ${err}`);
+function createOneResultPerPostcode(api_results) {
```

Comment:
@mlvernay This is where I took a decision that could be discussed. 

Now that we have an API that send multiple postcodes for one city we have 2 choices:

- Option 1: Send the list of postcodes to the backend and handle it there
- Option 2: Duplicate results and show one item in the auto-complete per postcode

I was going for option 1 when I met an issue. This means when looking for Paris it will look for all the postal codes (all arrondissements) at the same time, which makes the search pretty useless. (This is also true for smaller cities with multiples postcode like Lille but in this case I would consider it OK).



I went for option number 2 which adds a bit of JS here. I need more time to test, but from there the backend should do the job of filtering on the correct postcode. The main drawback of this is that the list in the autocomplete can be very long, see images

![Capture d’écran de 2023-06-20 11-27-07](https://github.com/MTES-MCT/acceslibre/assets/6287044/d7b16cae-af36-418b-8b7f-ee425b1a8d55)

![Capture d’écran de 2023-06-20 11-26-31](https://github.com/MTES-MCT/acceslibre/assets/6287044/446788f9-b11a-4c75-9326-12296b7ebaa0)

Reply:
Go for option 1.